src/WCSimWCTrigger.cc

```diff
@@ -0,0 +1,523 @@
+  // optionally removes digits from the input digits collection (when running different Alg* methods concurently) 
+  // so they are not used in subsequent trigger decisions or saved twice
+  //Also, only save digits of a specific type (again for when running different Alg* methods concurently)
+
+  // Add dummy triggers / exit without saving triggers as required
+  //
+  //saveFailuresMode = 0 - save only triggered events
+  //saveFailuresMode = 1 - save both triggered & not triggered events
+  //saveFailuresMode = 2 - save only not triggered events
+  if(TriggerTimes.size()) {
+    if(saveFailuresMode == 2)
+      return;
+  }
+  else {
+    if(saveFailuresMode == 0)
+      return;
+    TriggerTypes.push_back(kTriggerFailure);
+    TriggerTimes.push_back(saveFailuresTime);
+    TriggerInfos.push_back(std::vector<Float_t>(1, -1));
+    save_triggerType = kTriggerFailure;
+  }
+
+  //Get the PMT info for hit time smearing
+  G4String WCIDCollectionName = myDetector->GetIDCollectionName();
+  WCSimPMTObject * PMT = myDetector->GetPMTPointer(WCIDCollectionName);
+
+  //Loop over trigger times
+  for(unsigned int itrigger = 0; itrigger < TriggerTimes.size(); itrigger++) {
+    TriggerType_t triggertype = TriggerTypes[itrigger];
+    //check if we've already saved this trigger
+    if(triggertype != save_triggerType && save_triggerType != kTriggerUndefined)
+      continue;
+    float         triggertime = TriggerTimes[itrigger];
+    std::vector<Float_t> triggerinfo = TriggerInfos[itrigger];
+
+    //these are the boundary of the trigger gate: we want to add all digits within these bounds to the output collection
+    float lowerbound = triggertime + GetPreTriggerWindow(triggertype);
+    float upperbound = triggertime + GetPostTriggerWindow(triggertype);
+
+#ifdef WCSIMWCTRIGGER_VERBOSE
+    G4cout << "Saving trigger " << itrigger << " of type " << WCSimEnumerations::EnumAsString(triggertype)
+	   << " in time range [" << lowerbound << ", " << upperbound << "]"
+	   << " with trigger time " << triggertime
+	   << " and additional trigger info";
+    for(std::vector<Float_t>::iterator it = triggerinfo.begin(); it != triggerinfo.end(); ++it)
+      G4cout << " " << *it;
+    G4cout << G4endl;
+#endif
+
+    //loop over PMTs
+    for (G4int i = 0; i < WCDCPMT->entries(); i++) {
+      int tube=(*WCDCPMT)[i]->GetTubeID();
+      //loop over digits in this PMT
+      for ( G4int ip = 0; ip < (*WCDCPMT)[i]->GetTotalPe(); ip++){
+	int digit_time  = (*WCDCPMT)[i]->GetTime(ip);
+	if(digit_time >= lowerbound && digit_time <= upperbound) {
+	  //hit in event window
+	  //add it to DigitsCollection
+
+	  //first smear the charge & time
+	  float peSmeared = (*WCDCPMT)[i]->GetPe(ip);
+	  float Q = (peSmeared > 0.5) ? peSmeared : 0.5;
+	  G4double digihittime = -triggertime
+	    + WCSimWCTriggerBase::offset
+	    + digit_time
+	    + PMT->HitTimeSmearing(Q);
+	  if(digihittime < 0)
+	    continue;
```

Comment:
Ok, let's leave this for a future PR as well since it was also a problem in the old digitizer routine and doesn't really fit the mandate of this pull request. Make an issue on the tracker so we remember. I will have to think about why this continue was put in to begin with. I wonder if there is some situation where it is needed. We can discuss it more when the issue page gets created.